ukvm/ukvm_module_gdb.c

```diff
@@ -19,308 +19,206 @@
-
-    memset(&sockaddr, '\000', sizeof(sockaddr));
-    sockaddr.sin_family = AF_INET;
-    sockaddr.sin_port = htons(portn);
-    sockaddr.sin_addr.s_addr = htonl(INADDR_ANY);
-
-    r = bind(listen_socket_fd, (struct sockaddr *) &sockaddr,
-             sizeof(sockaddr));
-    if (r == -1)
-        perror("Failed to bind socket");
-
-    r = listen(listen_socket_fd, 0);
-    if (r == -1)
-        perror("Failed to listen on socket");
-
-    sockaddr_len = sizeof(sockaddr);
-    socket_fd = accept(listen_socket_fd, (struct sockaddr *) &sockaddr,
-                       &sockaddr_len);
-    if (socket_fd == -1)
-        perror("Failed to accept on socket");
+#error Unsupported target
 
-    close(listen_socket_fd);
+#endif 
 
-    protoent = getprotobyname("tcp");
-    if (!protoent) {
-        perror("getprotobyname (\"tcp\") failed");
-        return;
-    }
+static bool use_gdb = false;
+static int socket_fd = 0;
+static int portno = 1234; /* Default port number */
+static const unsigned char hexchars[] = "0123456789abcdef";
 
-    /* Disable Nagle - allow small packets to be sent without delay. */
-    opt = 1;
-    r = setsockopt(socket_fd, protoent->p_proto, TCP_NODELAY, &opt,
-                   sizeof(opt));
-    if (r == -1)
-        perror("setsockopt(TCP_NODELAY) failed");
+#define BUFMAX                         4096
+static unsigned char in_buffer[BUFMAX];
+static unsigned char registers[BUFMAX];
 
-    int ip = sockaddr.sin_addr.s_addr;
+/* The actual error code is ignored by GDB, so any number will do. */
+#define GDB_ERROR_MSG                  "E01"
 
-    printf("GDB Connected to %d.%d.%d.%d\n", ip & 0xff, (ip >> 8) & 0xff,
-           (ip >> 16) & 0xff, (ip >> 24) & 0xff);
+static int hex(unsigned char ch)
+{
+    if ((ch >= 'a') && (ch <= 'f'))
+        return (ch - 'a' + 10);
+    if ((ch >= '0') && (ch <= '9'))
+        return (ch - '0');
+    if ((ch >= 'A') && (ch <= 'F'))
+        return (ch - 'A' + 10);
+    return -1;
 }
 
-
-static char buf[4096], *bufptr = buf;
-static void flush_debug_buffer(void)
+/*
+ * Converts the memory pointed to by mem into an hex string in buf.
+ * Returns a pointer to the last char put in buf (null).
+ */
+static unsigned char *mem2hex(unsigned char *mem,
+                              unsigned char *buf, size_t count)
 {
-    char *p = buf;
-
-    while (p != bufptr) {
-        int n;
+    size_t i;
+    unsigned char ch;
 
-        n = send(socket_fd, p, bufptr - p, 0);
-        if (n == -1) {
-            perror("error on debug socket: %m");
-            break;
-        }
-        p += n;
+    for (i = 0; i < count; i++) {
+        ch = *mem++;
+        *buf++ = hexchars[ch >> 4];
+        *buf++ = hexchars[ch % 16];
     }
-    bufptr = buf;
-}
-
-
-void putDebugChar(int ch)
-{
-    if (bufptr == buf + sizeof(buf))
-        flush_debug_buffer();
-    *bufptr++ = ch;
+    *buf = 0;
+    return buf;
 }
 
-
-int getDebugChar(void)
+/*
+ * Converts the hex string in buf into binary in mem.
```

Comment:
In all of these cases `strlen(buf) == count`, however this function could be used for other commands where the data sent from GDB is made of multiple smaller commands.